### ISSUE TITLE
Block I/O tuning - need help/pointers to contribute this

### ISSUE THREAD

**Issue by zulh-civo** [kubevirt/kubevirt#5444]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> /kind enhancement

**What happened**:

I'm running a multi-tenant cluster where every tenant/user has Kubevirt VMs running in their own tenant network. There were some VMs that are using a lot of read IO — which causing high IO wait on the underlying node.

**What you expected to happen**:

AFAIK, there is no block I/O tuning support Kubevirt (yet). But when I looked at [Libvirt docs](https://libvirt.org/formatdomain.html), I saw this section:

> The optional iotune element provides the ability to provide additional per-device I/O tuning, with values that can vary for each device (contrast this to the `<blkiotune>` element, which applies globally to the domain). Currently, the only tuning available is Block I/O throttling for qemu. This element has optional sub-elements; any sub-element not specified or given with a value of 0 implies no limit. 

So I ran some tests using XMLs below and I can confirm Libvirt does support block I/O tuning.

Note:
The parameter values/numbers below are just examples. In real world, they are declarable via Kubevirt YAML file/Go struct.

XML 1:

```
<domain type='qemu' ...>
...
<devices>
    <emulator>/usr/local/bin/qemu-system-x86_64</emulator>
    <controller type='usb' model='ehci'/>
    <disk type='file' device='disk'>
        <driver name='qemu' type='qcow2'/>
        <source file='full/path/to/ubuntu.qcow2'/>
        <target dev='vda' bus='virtio'/>
        <iotune>
            <total_bytes_sec>52428800</total_bytes_sec>
            <total_iops_sec>1000</total_iops_sec>
        </iotune>
    </disk>
</devices>
...
</domain>
```

XML 2:

```
<domain type='qemu' ...>
...
<devices>
    <emulator>/usr/local/bin/qemu-system-x86_64</emulator>
    <controller type='usb' model='ehci'/>
    <disk type='file' device='disk'>
        <driver name='qemu' type='qcow2'/>
        <source file='full/path/to/ubuntu.qcow2'/>
        <target dev='vda' bus='virtio'/>
        <iotune>
            <read_bytes_sec>2000000</read_bytes_sec>
            <write_bytes_sec>1000000</write_bytes_sec>
            <read_iops_sec>100</read_iops_sec>
            <write_iops_sec>200</write_iops_sec>
        </iotune>
    </disk>
</devices>
...
</domain>
```

Now I'd like to add/expose those Libvirt parameters on Kubevirt side so I can create Kubevirt VM (using Go) with block I/O tuning support.

I've looked at the Kubevirt codebase and docs (this repository) and [kubevirt/client-go](https://github.com/kubevirt/client-go) repository to understand Kubevirt architecture and components but I'm lost and don't know where to start and what should I do to add support for parameters above.

I need help, pointers and suggestions from everyone here for me to contribute this to Kubevirt. FYI, I get how Kubernetes Operators work (I wrote a couple of k8s operators using Operator SDK) and I'm familiar with OSS/GitHub contribution concept — but I'm not that familiar with Kubevirt internals and that's what I need help with here.

To be more specific:

1. Which files and lines should I edit?
2. Do I need to run any generator/scripts during/after the edits?
3. Which Go test files and lines should I edit?
4. Do I need to edit anything in [kubevirt/client-go](https://github.com/kubevirt/client-go) repository? I believe I have to.
5. Where is the best place to add those parameters? Is inside `Disk: &kubevirtv1.DiskTarget{}` fine? Something like...

    ```
	Devices: kubevirtv1.Devices{
		AutoattachGraphicsDevice: &autottachGraphicsDevice,
		Disks: []kubevirtv1.Disk{
			{
				Name: "containerdisk",
				DiskDevice: kubevirtv1.DiskDevice{
					Disk: &kubevirtv1.DiskTarget{
						Bus: "virtio",
                                            // here is fine?
					},
				},
			}
		},
	},
    ```

**How to reproduce it (as minimally and precisely as possible)**:

N/A

**Anything else we need to know?**:

External resources that I found when researching about this topic:

* https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/virtualization_tuning_and_optimization_guide/sect-virtualization_tuning_optimization_guide-blockio-techniques
* https://docs.fedoraproject.org/en-US/Fedora_Draft_Documentation/0.1/html/Virtualization_Deployment_and_Administration_Guide/sect-libvirt-dom-xml-blk-io-tuning.html

**Environment**:
- KubeVirt version (use `virtctl version`): N/A
- Kubernetes version (use `kubectl version`): N/A
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

**Comment by zulh-civo** [comment]
@phoracek

**Comment by phoracek** [comment]
You mention that you tried using libvirt to confirm whether it really has that attribute. What I'm wondering is whether you tried it to prove that setting it resolves the I/O issue on the cluster, with realistic workload.

**Comment by alicefr** [comment]
@zulh-civo the current issue is that you should specify those values for a pvc, and then kubevirt should read them, and create the xml option accordingly. Currently, the k8s APIs don't have a field for this, the workaround is to use a custom annotation that kubevirt could read. The real question is if the iotune should be exposed to the user or not, and how you generate those values.

If you want to look at an example that checks for annotation and set some xml option, you can check [this commit](https://github.com/kubevirt/kubevirt/pull/5006/commits/4b2c0eb68bf8786951544d6d4bdee6b6fa5f791b)

**Comment by aglitke** [comment]
@zulh-civo I see in your initial comment on the PR you use a containerDisk as your example.  Then the conversation moved to PVCs.  While PVCs are likely the primary use case for iotune I'm not sure about limiting it to PVCs by placing annotations on PVCs and DataVolumes.  Semantically, iotune (as implemented) is a directive to the workload not a property of the storage.  To me it makes more sense to hang this API on the VMI/VM.  This would allow you to use it with other volume types such as containerDisk.

**Comment by zulh-civo** [comment]
@aglitke do you have any recommendation where the iotune annotations should be placed if it's not inside PVC/DataVolume? 

Or, are you saying that we shouldn't use annotations at all and add new struct fields to VMI/VM instead? If that's what you meant, where is the best place to put it?

(by asking "where" above, I'd really appreciate it if you can point me out to the specific file(s) and line number(s) - I'm very new to this project)

If we place this iotune fields to VMI/VM, does it mean all attached storage (e.g. cdrom, floppy & etc) XML will have the translated/converted iotune XML as well? Help me to figure out the usage part of this after the declaration in the VMI/VM. Maybe you can share some YAMLs?

**Comment by mlsorensen** [comment]
I commented on this in the groups thread but just for the sake of tracking on the issue - There is a generic way provided by Kubevirt to mutate the VM definition - have you considered making a simple sidecar hook to inject the iotune into your VMs? That would be an even faster solution that you'd have complete control of.

You'd create a docker image with a small golang program that can edit the libvirt XML of the VM. The selection of this image is controlled by annotation on the VM:

```
annotations:
  hooks.kubevirt.io/hookSidecars: '[{"image": "example-hook-sidecar:devel"}]'
  iotune.my.org/iops.read: 1000
```

Kubevirt calls this image and lets it do some prep work. The [community example ](https://github.com/kubevirt/kubevirt/blob/master/cmd/example-hook-sidecar/smbios.go)sets the baseboard manufacturer.

**Comment by zulh-civo** [comment]
@mlsorensen thank you for sharing that. I didn't know Kubevirt has hooks support. If I choose this approach, will it work even if the [schema](https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-launcher/virtwrap/api/schema.go) doesn't have iotune fields declared?

In my PR, I added the iotune fields to the libvirt schema ([see here](https://github.com/kubevirt/kubevirt/pull/5573/files#diff-66ca092e3693da97e42a13a63ea9d298df19057a358558314e23097bc32c3035)). I wonder if I should create another PR to only add those iotune fields?

**Comment by davidvossel** [comment]
> In my PR, I added the iotune fields to the libvirt schema (see here). I wonder if I should create another PR to only add those iotune fields?


My expectation is that this will work. 

The sidecar hooks work exclusively with xml. So as long as your hook consistently injects that field into the xml every time the hook is called, then your domain changes should stick around and be persisted into libvirt. 

if this doesn't work for you, then it's likely a bug in our sidecar hooks implementation.
